DiverseSelector/base.py

```diff
@@ -136,59 +179,41 @@ def __init__(self,
         #                          f"than {self.enhanced_sampling_weight} of number of selected"
         #                          f"molecules {self.num_selected_mols}.")
 
-    # def enhanced_sampling(self):
-    #     """Enhanced sampling with random sampling or Monte Carlo sampling."""
-    #     pass
+    def cluster(self, clustering_method:str):
+        """
+        Perform the clastering of the dataset
+        :param clustering_method: selected method for clustering (k-means, DBSCAN etc)
+        :return:
+        """
+        pass
 
-    @abstractmethod
-    def select(self):
-        """Add support of clustering and enhanced sampling."""
+    def select(self, num_selected: int, random_seed=None):
+        """
+        Select the subset molecules with optimal diversity.
+        :param num_selected: amount of molecule that need to be selected
+        :param random_seed: int
+        :return:
+        """
+        self.random_seed = 42 if random_seed is None else random_seed
+        # check if number of clusters is less than number of selected molecules
+        if not self.num_clusters > num_selected:
+            raise ValueError("The number of clusters is great than number of selected molecules.")
+        # check if we have valid number of clusters because selecting 1.5 molecules is not practical
+        if int(num_selected / num_clusters) - num_selected / num_clusters != 0:
+            raise ValueError("The number of molecules in each cluster should be an integer.")
         pass
```

Comment:
Why don't we have the abstract method for `ClusteringSelectionBase`?

Reply:
Same explanation as for the `DissimilaritySelectionBase` class.

Overall I see the structure in this way:
1. Every selector in as _concrete_ class which is group tools for the "family" of similar approaches and methods: ClusterSelector, DissimilaritySelector, MonteCarlo selector. It's obvious, that these classes should have similar structure and some common methods (IO for example), so how to define the behaviour of this classes?
2. Behaviour, structure and common methods for these classes is described and defined by an abstract class called `Selector`

You can think about it as a boxes. Each Class is a box with different tools that are related to some approach(Dissimilarity, Clustering etc). However, in order to force everyone to create boxes of similar size/ weight/color and amount of tools there is a description, blueprint of the box -- abstract class called Selector

Hope it clarifies what points I tried to address by changing the existing hierarchy

Reply:
Thanks for the detailed explanations! I originally followed a similar pattern in the very first initial commits, but later I realized that this is not an easy thing. But let's leave it as it is at this moment.